### ISSUE TITLE
Tinker compatibility: missing components inventory (gap list for #42)

### ISSUE THREAD

**Issue by sfc-gh-kganesan** [Snowflake-AI-Research/Arctic-Platform#90]
Companion to #42, which holds the design and the "no surprises" bar. This issue is
the **buildable inventory**: every component still missing, which repo it lands in,
and the evidence for why it's needed. Prior art is #43 (Tinker HTTP layer v1), which
proved the on-prem path.

Two things motivated splitting this out. Investigating Jaesong's Cortex recipes
(`cortex-client/recipes/`) and the ArcticTraining pipeline turned up **three gaps not
in #42**, and it **retired one blocker** that #42 and I had both mis-scoped.

---

## Already proven — do not re-litigate

Worth stating explicitly so these stop being treated as open risks.

| Capability | Evidence |
| :--- | :--- |
| Cookbook RL recipe runs on the Tinker wire | #43: `recipes/rl/tinker/grpo_e2e.sh` drives unmodified `tinker_cookbook.recipes.math_rl.train`, mirroring SkyRL-tx's `gsm8k_tinker.sh` CI (cookbook @ `016468b0f2`). FFT only. |
| Per-token logprobs come back from `forward_backward` | #43 requests `post: ["compute_entropy_and_logprobs"]` and reads `batch.logprobs`; `compute_logprobs` / `compute_entropy_and_logprobs` are registered post-processors (`pipeline.py:727,790`) and `ray_server.py:783-794` merges them. |
| LoRA r=32 on Cortex | `recipes/rl_loop.py:263-272` ships a real `peft_config` (`peft_type: Lora`, `r == lora_alpha == 32`) on both sub-jobs; recipes converge. |
| Per-step learning-rate override | `neutrino_common.py:247` → `client.step(job_id, learning_rate=...)`. |
| Sampling + weight sync with LoRA | `rl_loop.py:554` syncs with `weight_format="lora"`. |
| `grpo` as a server-side loss | `rl_loop.py:349` `loss_fn="grpo"`; registered at `grpo.py:236`. |

## Correction: logprobs are not a Cortex ask

The DSS spec (§6.1) warns that `post_process_outputs` is returned as an empty object
and that callers "must not assume that `compute_logprobs` appears there." That is
true and it points at the wrong key — the pipeline puts post-processor results under
`batch`:

```python
# ArcticTraining .../arctic_rl/processors/pipeline.py:402-406
model_outputs.update(**post_process_outputs)

pipeline_outputs = dict(
    batch=post_process_outputs,
    metrics=dict(),
)
```

Cortex runs this same pipeline, and #43 already consumes `batch.logprobs` successfully
on-prem. So this is a **confirmation task on QA6**, not a cross-team feature request.
Retiring it as a blocker.

---

### A. Weighted cross-entropy — ArcticTraining (DSS fork)

The single most load-bearing gap. `register_loss_fn` has exactly two entries,
`grpo` (`grpo.py:236`) and `verl_grpo` (`verl_grpo.py:321`). There is **no
`cross_entropy` loss function**. CE on Cortex today comes from HuggingFace's own
`labels` path — `sft_loop.py:214-222` passes `labels` in `kwargs` and no `processing`
block at all, so the model computes CE internally with `-100` masking.

That is binary keep/drop masking. Tinker's `cross_entropy` takes **real-valued
per-token weights**, and `forward_backward_custom` is implemented client-side by
compiling any custom loss into weighted CE (`weights = -dC/dlogprobs`, exploiting
`L = sum(-logprobs * weights)` being linear in logprobs). A `-100` label can say
"skip this token"; it cannot say "weight this token by 0.37".

- [ ] `@register_loss_fn("cross_entropy")` honouring float per-token weights and Tinker's reduction
- [ ] Unit test: recovers a locally computed gradient for a non-binary weight vector

Blocks: all supervised recipes, and every Phase 2 custom loss (DPO, SDFT, GKD, OPD).

### B. Per-backend datum encoding — `arctic_platform/tinker/`

Cortex and on-prem do not take the same batch shape. Cortex is RPC-style
(`neutrino_common.py:236-240`):

```python
{"args": (), "kwargs": {input_ids, attention_mask, position_ids, use_cache, labels?},
 "context": {input_ids, advantages, loss_mask},
 "processing": {loss_fn, config, post}}
```

whereas the on-prem/verl path sends `{"batch": ..., "meta": ...}`. The DSS spec (§10.3)
is explicit that `context` and `processing` are open backend contracts. Today
`datum_list_to_arctic_batch` emits only the verl shape.

- [ ] Per-backend datum encoder behind one adapter seam (the concrete form of the existing `TODO(unify-backends)` in `client/requests.py`)
- [ ] Parity test: same `Datum` list, both encoders, equal loss

### C. Fail loud instead of empty logprobs — `arctic_platform/tinker/router.py`

When Arctic returns no logprobs the router substitutes empty dicts (`router.py:962-968`).
Cookbook then does an unconditional `output["logprobs"]`
(`tinker_cookbook/rl/train.py:277-280`), so the user gets a bare `KeyError` mid-run.

Stubbing is not an option either: those logprobs feed `compute_kl_sample_train`, the
sampler-vs-trainer divergence metric — the exact alarm that catches the class of bug a
new backend integration is most likely to introduce. Fabricating it disables the
alarm for the fire we're most likely to start.

- [ ] Remove the silent fallback; refuse at boot via capability negotiation naming the missing post-processor

### D. Client timeout / cold start — `arctic_platform/tinker/`

#43 had to monkey-patch the SDK's 60s HTTP timeout to 1800s
(`examples_math_rl_driver.py`) because multi-GPU ZeRO-3 forward+backward exceeds it.
A stock `pip install tinker` user has no such patch, so this is a first-run surprise.
Cortex compounds it: submit-then-poll with a 0.5s default poll interval, plus
`job_ready_timeout` defaulting to 1800s for provisioning.

- [ ] Keep every verb inside the stock SDK timeout by returning futures promptly and doing work asynchronously (ties to the `create_model` / future-store item on #42)
- [ ] Measure per-verb latency against Cortex and record the budget

### E. Cortex transport gaps — `arctic_platform/client/transports/cortex.py`

- [ ] sleep/wake ops as no-ops — they raise `NotImplementedError` (`cortex.py:246`) while `AsyncArcticClient.sync_weights` calls `wake_inference` first (`base.py:271`), so **weight sync fails on its first call against Cortex**
- [ ] `forward` endpoint (Phase 2 gate; currently `NotImplementedError`)

### F. Verification on QA6 — needs one GPU job

- [ ] Confirm `batch.logprobs` is populated on Cortex when `post: ["compute_entropy_and_logprobs"]` is requested. Minimal delta on a known-good path: reuse `sft_loop.py`'s `job_body` (training-only sub-job, `flash_attention_3`, bf16, ZeRO-2, LoRA-32) and add the one processing field.
- [ ] Then: one real cookbook recipe end to end against Cortex.

---

## Everything else

The remaining Phase 1 items — `AdamParams`, caller-owned grad accumulation via
`managed_gradient_accumulation: false`, on-demand provisioning, model registry keyed
by `model_id`, `unload_model`, `save_state`/`load_state`, async sampling snapshots,
boot-time capability negotiation — are tracked on #42 and unchanged. Lifting #43's
forced `lora_rank=0` ("Arctic v1 = FFT only") is also on #42; item (A) above is the
newly-identified prerequisite that was hiding behind it.

## Sequencing

1. Land the #43 re-target onto the unified client (in progress) and get the suite green.
2. (F) — one job, retires the last unknown.
3. (A) weighted CE, then (B) the Cortex encoder. These two unlock the `sl` profile on Cortex.
4. (C), (D), (E) — the no-surprises hardening.
5. Phase 2 `forward`, per #42.

## Known cost, not a gap

`forward_backward_custom` is inherently two passes — a `forward` for logprobs, then a
`forward_backward` with the derived weights — so every custom-loss recipe pays double
the forward compute and batch upload per step. That's upstream Tinker's design, not
something we introduce. Bandwidth is not the concern: logprobs are `[B, S]` float32,
about half the bytes of the `input_ids` already uploaded, and ~150,000x smaller than
the `[B, S, V]` logits we deliberately never put on the wire.
